File: python/deep_tutor/agent/agent.py
"""
Deep Tutor Agent

Tutor agent using create_deep_agent with FilesystemBackend.

Personality and teaching guidelines are loaded from AGENTS.md via MemoryMiddleware.
Module skills are loaded from /skills/ via SkillsMiddleware (progressive disclosure).
The dynamic prompt injects per-session context: student name, module, and goals.

Assistant context (set per assistant in LangSmith):
- module_id:       which module to load (e.g. "module-1")
- store_namespace: student's Store namespace (derived from email, e.g. "jane@example_com")
"""

import asyncio
from dataclasses import dataclass
from pathlib import Path

from deepagents import create_deep_agent
from deepagents.backends.filesystem import FilesystemBackend
from deepagents.middleware.permissions import FilesystemPermission
from langchain.agents.middleware import ModelRequest, dynamic_prompt
from langchain.chat_models import init_chat_model
from langchain_mcp_adapters.client import MultiServerMCPClient

# ---------------------------------------------------------------------------
# Model
# ---------------------------------------------------------------------------
model = init_chat_model("anthropic:claude-sonnet-4-6")
# model = init_chat_model("openai:gpt-4o")
# model = init_chat_model("google_genai:gemini-2.5-flash")


# ---------------------------------------------------------------------------
# Context schema
# ---------------------------------------------------------------------------
@dataclass
class ContextSchema:
    module_id: str = "module-1"
    store_namespace: str = ""  # derived from student email


# ---------------------------------------------------------------------------
# Dynamic system prompt
# ---------------------------------------------------------------------------
@dynamic_prompt
async def tutor_prompt(request: ModelRequest) -> str:
    module_id = request.runtime.context.module_id
    store_namespace = request.runtime.context.store_namespace
    student_name = "Student"
    goals = ""

    if store_namespace and request.runtime.store:
        item = await request.runtime.store.aget((store_namespace,), key="profile")
        if item is not None:
            profile = item.value
            student_name = f"{profile.get('first_name', '')} {profile.get('last_name', '')}".strip() or "Student"
            goals = profile.get("goals", "")

    goals_line = f"\n\nThe student's goals: {goals}" if goals else ""
    return (
        f"Your student's name is {student_name}. "
        f"They are working on {module_id}.{goals_line}\n\n"
        f"On your first response, read /skills/{module_id}/SKILL.md. "
        f"That file contains your teaching instructions — follow them. "
        f"When you need to answer a factual question, read /skills/{module_id}/information.md."
    )


# ---------------------------------------------------------------------------
# MCP tools
# ---------------------------------------------------------------------------
_mcp_client = MultiServerMCPClient({
    "docs": {
        "url": "https://docs.langchain.com/mcp",
        "transport": "http",
    }
})
mcp_tools = asyncio.run(_mcp_client.get_tools())


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------
backend = FilesystemBackend(
    root_dir=Path(__file__).parent.parent,
    virtual_mode=True,
)

graph = create_deep_agent(
    model=model,
    tools=[*mcp_tools],
    middleware=[tutor_prompt],
    backend=backend,
    memory=["/AGENTS.md"],
    skills=["/skills/"],
    permissions=[
        FilesystemPermission(operations=["write"], paths=["/skills/**"], mode="deny"),
    ],
    context_schema=ContextSchema,
)

File: typescript/m5/l5/agent/agent.ts
/**
 * Deep Tutor Agent — TypeScript
 *
 * Tutor agent using createDeepAgent with FilesystemBackend.
 *
 * Personality and teaching guidelines are loaded from AGENTS.md via MemoryMiddleware.
 * Module skills are loaded from /skills/ via SkillsMiddleware (progressive disclosure).
 * A custom middleware injects per-session context: student name, module, and goals.
 *
 * Assistant context (set per assistant in LangSmith):
 * - module_id:       which module to load (e.g. "module-1")
 * - store_namespace: student's Store namespace (derived from email, e.g. "jane@example_com")
 */

import path from "node:path";
import { fileURLToPath } from "node:url";
import { z } from "zod";

import { createDeepAgent, FilesystemBackend } from "deepagents";
import { createMiddleware } from "langchain";
import { getConfig, getStore } from "@langchain/langgraph";
// import { ChatGoogleGenerativeAI } from "@langchain/google-genai";  // NOTE: Gemini 2.5 Flash returns empty output with multi-block system messages — use Anthropic or OpenAI
// import { ChatOpenAI } from "@langchain/openai";
import { ChatAnthropic } from "@langchain/anthropic";
import { MultiServerMCPClient } from "@langchain/mcp-adapters";

// ---------------------------------------------------------------------------
// Model
// ---------------------------------------------------------------------------
const model = new ChatAnthropic({ model: "claude-sonnet-4-6" });
// const model = new ChatOpenAI({ model: "gpt-4o" });
// const model = new ChatGoogleGenerativeAI({ model: "gemini-2.5-flash" }); // see note above

// ---------------------------------------------------------------------------
// Context schema
// ---------------------------------------------------------------------------
const contextSchema = z.object({
  module_id: z.string().default("module-1"),
  store_namespace: z.string().default(""), // derived from student email
});

// ---------------------------------------------------------------------------
// Dynamic system prompt middleware
//
// Uses createMiddleware + wrapModelCall to APPEND to the system message
// rather than replace it, so memory and skills layers built by createDeepAgent
// are preserved.
// ---------------------------------------------------------------------------
const tutorPrompt = createMiddleware({
  name: "TutorPrompt",

  async wrapModelCall(request: any, handler: any) {
    const ctx = (getConfig().context ?? {}) as Record<string, string>;
    const moduleId = ctx.module_id ?? "module-1";
    const storeNamespace = ctx.store_namespace ?? "";
    let studentName = "Student";
    let goals = "";

    if (storeNamespace) {
      const store = getStore();
      if (store) {
        const item = await store.get([storeNamespace], "profile");
        if (item) {
          const profile = item.value as Record<string, string>;
          const fullName = `${profile.first_name ?? ""} ${profile.last_name ?? ""}`.trim();
          if (fullName) studentName = fullName;
          if (profile.goals) goals = profile.goals;
        }
      }
    }

    const goalsLine = goals ? `\n\nThe student's goals: ${goals}` : "";
    const promptText =
      `Your student's name is ${studentName}. ` +
      `They are working on ${moduleId}.${goalsLine}\n\n` +
      `On your first response, read /skills/${moduleId}/SKILL.md. ` +
      `That file contains your teaching instructions — follow them. ` +
      `When you need to answer a factual question, read /skills/${moduleId}/information.md.`;

    return handler({
      ...request,
      systemMessage: request.systemMessage.concat(promptText),
    });
  },
});

// ---------------------------------------------------------------------------
// MCP tools
// ---------------------------------------------------------------------------
const mcpClient = new MultiServerMCPClient({
  docs: {
    url: "https://docs.langchain.com/mcp",
    transport: "http",
  },
});

const mcpTools = await mcpClient.getTools();

// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const backend = new FilesystemBackend({
  rootDir: path.join(__dirname, ".."),
  virtualMode: true,
});

export const graph = await createDeepAgent({
  model,
  tools: mcpTools,
  middleware: [tutorPrompt],
  backend,
  memory: ["/AGENTS.md"],
  skills: ["/skills/"],
  permissions: [
    { operations: ["write"], paths: ["/skills/**"], mode: "deny" },
  ],
  contextSchema,
});
